pkg/services/grafana-apiserver/rest/dualwriter.go

```diff
@@ -0,0 +1,117 @@
+package rest
+
+import (
+	"context"
+
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+var (
+	_ rest.Storage              = (*DualWriter)(nil)
+	_ rest.Scoper               = (*DualWriter)(nil)
+	_ rest.TableConvertor       = (*DualWriter)(nil)
+	_ rest.CreaterUpdater       = (*DualWriter)(nil)
+	_ rest.CollectionDeleter    = (*DualWriter)(nil)
+	_ rest.GracefulDeleter      = (*DualWriter)(nil)
+	_ rest.SingularNameProvider = (*DualWriter)(nil)
+)
+
+// UnifiedStorage is a storage implementation that writes to the Grafana Unified Storage.
+type UnifiedStorage interface {
+	rest.Storage
+	rest.StandardStorage
+	rest.Scoper
+	rest.TableConvertor
+	rest.SingularNameProvider
+}
+
+// SQLStorage is a storage implementation that writes to the Grafana SQL database.
+type SQLStorage interface {
+	rest.Storage
+	rest.Scoper
+	rest.SingularNameProvider
+	rest.TableConvertor
+}
+
+// DualWriter is a storage implementation that writes first to SQLStorage and then to UnifiedStorage.
+// If writing to SQLStorage fails, the write to UnifiedStorage is skipped and the error is returned.
+// UnifiedStorage is used for all read operations.
+//
+// The SQLStorage implementation must implement the following interfaces:
+// - rest.Storage
+// - rest.TableConvertor
+// - rest.Scoper
+// - rest.SingularNameProvider
+//
+// These interfaces are optional, but they all should be implemented to fully support dual writes:
+// - rest.Creater
+// - rest.Updater
+// - rest.GracefulDeleter
+// - rest.CollectionDeleter
```

Comment:
not sure if this is a good idea, but by making these interfaces optional we can limit the amount of verbs we expose via discovery when dual write is disabled.

for example, with dual write disabled:
```
kubectl get --raw="/apis/playlist.x.grafana.com/v0-alpha" | jq
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "playlist.x.grafana.com/v0-alpha",
  "resources": [
    {
      "name": "playlists",
      "singularName": "playlist",
      "namespaced": true,
      "kind": "Playlist",
      "verbs": [
        "get",
        "list"
      ]
    }
  ]
}
```

with dual write enabled:
```
kubectl get --raw="/apis/playlist.x.grafana.com/v0-alpha" | jq
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "playlist.x.grafana.com/v0-alpha",
  "resources": [
    {
      "name": "playlists",
      "singularName": "playlist",
      "namespaced": true,
      "kind": "Playlist",
      "verbs": [
        "create",
        "delete",
        "deletecollection",
        "get",
        "list",
        "patch",
        "update",
        "watch"
      ]
    }
  ]
}
```

Reply:
for this to work we will really need everything except "watch" and "deletecollection" in the legacy storage, but I think this is a good idea